### PR TITLE
ddl: ddl-owner try to use memory infoSchema to check first

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
Make ddl-owner try to use memory infoschema to check first if the owner tidb-server already has loaded the latest schema.

Here is a simple test, There are 11394 tables in test20.
Deploy  1 tidb, 1 pd, 1 tikv. 

## create table 
```
mysql root@localhost:test20> show tables;
11394 rows in set
Time: 0.253s
```

Before, spend 0.37s, `checkTableNotExists` spend 0.34s
```
mysql root@localhost:test20> create table tb (a int);
Query OK, 0 rows affected
Time: 0.371s 
```
In this PR:  spend 0.07s, `checkTableNotExists` spend 267us.
```
mysql root@localhost:test20> create table ta (a int);
Query OK, 0 rows affected
Time: 0.070s
```
## Create database

There are 18500 databases already in TiDB.

```sql
mysql root@localhost:test> show databases;
18500 rows in set
Time: 0.546s
```
Before: spend 0.32S, and checkSchemaExist spend 250ms.
```sql
mysql root@localhost:test> create database dbd4
Query OK, 0 rows affected
Time: 0.326s
```

In this PR: 0.08s, checkSchemaExist spend 1 ms.
```sql
mysql root@localhost:test> create database dbd2
Query OK, 0 rows affected
Time: 0.080s
```
### What is changed and how it works?
Change `checkTableNotExists` function. 

### Check List 

Tests 

 - No test...

Code changes

 - Has exported function/method change

Related changes

 - Need to cherry-pick to the release branch ?

### CODE REVIEW

File: ddl/table.go [hunk 1]
```diff
@@ -658,7 +658,32 @@ func onModifyTableCharsetAndCollate(t *meta.Meta, job *model.Job) (ver int64, _
 	return ver, nil
 }
 
-func checkTableNotExists(t *meta.Meta, job *model.Job, schemaID int64, tableName string) error {
+func checkTableNotExists(d *ddlCtx, t *meta.Meta, job *model.Job, schemaID int64, tableName string) error {
+	// d.infoHandle maybe nil in some test.
+	if d.infoHandle != nil {
```
Comment (winkyao) [inline ddl/table.go]:
```
if d.infoHandle != nil {
	return checkTableNotExistsFromStore(t, job, schemaID, tableName)
}

// Try to use memory schema info to check first.
		currVer, err := t.GetSchemaVersion()
		if err != nil {
			return err
		}
		is := d.infoHandle.Get()
		if is.SchemaMetaVersion() == currVer {
			// Check this table's database.
			schema, ok := is.SchemaByID(schemaID)
			if !ok {
				return infoschema.ErrDatabaseNotExists.GenWithStackByArgs("")
			}
			if is.TableExists(schema.Name, model.NewCIStr(tableName)) {
				return infoschema.ErrTableExists.GenWithStackByArgs(tableName)
			}
			return nil
		}
```

File: ddl/table.go [hunk 2]
```diff
@@ -658,7 +658,32 @@ func onModifyTableCharsetAndCollate(t *meta.Meta, job *model.Job) (ver int64, _
 	return ver, nil
 }
 
-func checkTableNotExists(t *meta.Meta, job *model.Job, schemaID int64, tableName string) error {
+func checkTableNotExists(d *ddlCtx, t *meta.Meta, job *model.Job, schemaID int64, tableName string) error {
+	// d.infoHandle maybe nil in some test.
```
Comment (winkyao) [inline ddl/table.go]:
```
if d.infoHandle == nil {
	return checkTableNotExistsFromStore(t, job, schemaID, tableName)
}

// d.infoHandle maybe nil in some test.
	
		// Try to use memory schema info to check first.
		currVer, err := t.GetSchemaVersion()
		if err != nil {
			return err
		}
		is := d.infoHandle.Get()
		if is.SchemaMetaVersion() == currVer {
			// Check this table's database.
			schema, ok := is.SchemaByID(schemaID)
			if !ok {
				return infoschema.ErrDatabaseNotExists.GenWithStackByArgs("")
			}
			if is.TableExists(schema.Name, model.NewCIStr(tableName)) {
				return infoschema.ErrTableExists.GenWithStackByArgs(tableName)
			}
			return nil
		}
	
 	return checkTableNotExistsFromStore(t, job, schemaID, tableName)
```

File: ddl/table.go [hunk 3]
```diff
@@ -658,7 +658,33 @@ func onModifyTableCharsetAndCollate(t *meta.Meta, job *model.Job) (ver int64, _
...
+	if d.infoHandle == nil {
+		return checkTableNotExistsFromStore(t, job, schemaID, tableName)
+	}
+	// Try to use memory schema info to check first.
+	currVer, err := t.GetSchemaVersion()
+	if err != nil {
+		return err
+	}
+	is := d.infoHandle.Get()
+	if is.SchemaMetaVersion() == currVer {
```
Comment (winkyao) [inline ddl/table.go]:
```
if is.SchemaMetaVersion() != currVer {
	return checkTableNotExistsFromStore(t, job, schemaID, tableName)
}

// Check this table's database.
schema, ok := is.SchemaByID(schemaID)
if !ok {
	return infoschema.ErrDatabaseNotExists.GenWithStackByArgs("")
}
if is.TableExists(schema.Name, model.NewCIStr(tableName)) {
	return infoschema.ErrTableExists.GenWithStackByArgs(tableName)
}
return nil
```

Comment (crazycs520) [reply ddl/table.go]: done.

**Comment by bb7133** [comment]
LGTM

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**